File: src/schedule-algorithm/algorithm.py
# Algorithm by Rohan "RJ" Jaiswal
# Adapted from chaudhary1337's solution on Leetcode 1494

import json
from collections import defaultdict

with open("cse.json", "r") as json_file:
    cse_json = json.load(json_file)

graph = {}
visited = {}
schedule = []

for course_name, course_data in cse_json.items():
    # Access the dependencies list
    dependencies = course_data.get("dependencies", [])
    # Populate the graph dictionary
    graph[course_name] = dependencies
    visited[course_name] = 0


def credit_sum(nodes: dict, cse_json: dict):
    return sum(cse_json[course]["credits"] for course in nodes)


in_degrees = defaultdict(int)
for course, dependencies in graph.items():
    for row in dependencies:
        for prereq in row:
            in_degrees[prereq] += 1


def recurse(graph, visited, in_degrees, credit_max, course_json):
    if not any(value > 0 for value in in_degrees.values()):
        return

    # select nodes that haven't been visited and don't have any dependencies
    nodes = {course: dependencies for course, dependencies in graph.items(
    ) if visited[course] == 0 and in_degrees[course] == 0}

    selected = {}
    while len(nodes) > 0 and credit_sum(selected, course_json) < 18:
        course, dependency = nodes.popitem()
        # if hours don't exceed 18, then add the course
        proposed_hours = credit_sum(
            selected, course_json) + course_json[course]["credits"]
        if proposed_hours > 18:
            break
        selected[course] = dependency

    semester = []
    for course, dependencies in selected.items():
        semester.append(course)
        # mark as visited
        visited[course] = 1
        # update the in-degrees
        for row in dependencies:
            for course in row:
                in_degrees[course] -= 1
    schedule.insert(0, semester)
    recurse(graph, visited, in_degrees, credit_max, course_json)


recurse(graph, visited, in_degrees, 18, cse_json)
print(schedule)

File: src/schedule-algorithm/algorithm.mjs
/** @format */

// Algorithm by Rohan "RJ" Jaiswal
// Adapted from chaudhary1337's solution on Leetcode 1494

export function convertToDict(nodes) {
  let result = {};

  for (let node of nodes) {
    let [department, classNumber] = node.id.split("-");

    let dependencies = node.dependencies.map((depGroup) =>
      depGroup.map((dep) => dep)
    );

    result[node.id] = {
      classNumber,
      department,
      name: node.title,
      dependencies: dependencies,
      credits: node.credits,
    };
  }

  return result;
}

export function generateSchedule(json, maxCredit) {
  const graph = {};
  const visited = {};
  const schedule = [];

  for (const [courseName, courseData] of Object.entries(json)) {
    const dependencies = courseData.dependencies || [];
    graph[courseName] = dependencies;
    visited[courseName] = 0;
  }

  const creditSum = (nodes, courseJson) => {
    return nodes.reduce((acc, course) => acc + courseJson[course].credits, 0);
  };

  const inDegrees = {};
  for (const [course, dependencies] of Object.entries(graph)) {
    for (const row of dependencies) {
      for (const prereq of row) {
        inDegrees[prereq] = (inDegrees[prereq] || 0) + 1;
      }
    }
  }

  const recurse = (graph, visited, inDegrees, creditMax, courseJson) => {
    if (!Object.values(inDegrees).some((value) => value > 0)) return;

    const nodes = Object.fromEntries(
      Object.entries(graph).filter(
        ([course]) => visited[course] === 0 && (inDegrees[course] || 0) === 0
      )
    );

    const selected = {};
    while (
      Object.keys(nodes).length > 0 &&
      creditSum(Object.keys(selected), courseJson) < 18
    ) {
      const course = Object.keys(nodes).pop();
      const dependencies = nodes[course];
      delete nodes[course];

      const proposedHours =
        creditSum(Object.keys(selected), courseJson) +
        courseJson[course].credits;
      if (proposedHours > creditMax) break;
      selected[course] = dependencies;
    }

    const semester = [];
    for (const [course, dependencies] of Object.entries(selected)) {
      semester.push(course);
      visited[course] = 1;
      for (const row of dependencies) {
        for (const course of row) {
          inDegrees[course] -= 1;
        }
      }
    }
    schedule.unshift(semester);
    recurse(graph, visited, inDegrees, creditMax, courseJson);
  };

  recurse(graph, visited, inDegrees, maxCredit, json);

  return schedule;
}

export function topologicalSort(data) {
  let visited = {};
  let stack = [];
  // let order = {};

  // Create a list of visited nodes initialized to false
  for (let course in data) {
    visited[course] = false;
  }

  // Utility function for the depth-first traversal of the graph
  function dfs(node) {
    visited[node] = true;

    let dependencies;
    try {
      dependencies = data[node].dependencies;
    } catch (error) {
      console.log("Error for node: " + node);
    }
    for (let i = 0; i < dependencies.length; i++) {
      for (let j = 0; j < dependencies[i].length; j++) {
        if (!visited[dependencies[i][j]]) {
          dfs(dependencies[i][j]);
        }
      }
    }

    // As we backtrack from recursion, push nodes to stack
    stack.push(node);
  }

  // Go through all nodes and perform depth-first traversal
  for (let course in data) {
    if (!visited[course]) {
      dfs(course);
    }
  }

  // while (stack.length) {
  //   order.push(stack.pop());
  // }

  return stack;
  // return order;
}

/**
 * Use cases:
 */
// const newFormats = convertToNewFormat(nodes);
// console.log(newFormats);

// const schedule = generateSchedule(newFormats, 18);
// console.log(schedule);
